cpp/ql/src/Likely Bugs/Conversion/CastArrayPointerArithmetic.ql

```diff
@@ -20,21 +20,29 @@ import cpp
 import semmle.code.cpp.dataflow.DataFlow
 import DataFlow::PathGraph
 
+Type getFullyConvertedType(DataFlow::Node node) {
+  result = node.asExpr().getFullyConverted().getUnspecifiedType()
+}
+
 class CastToPointerArithFlow extends DataFlow::Configuration {
   CastToPointerArithFlow() { this = "CastToPointerArithFlow" }
 
-  override predicate isSource(DataFlow::Node node) {
+  override predicate isSource(DataFlow::Node node, DataFlow::FlowState state) {
     not node.asExpr() instanceof Conversion and
     exists(Type baseType1, Type baseType2 |
       hasBaseType(node.asExpr(), baseType1) and
       hasBaseType(node.asExpr().getConversion*(), baseType2) and
       introducesNewField(baseType1, baseType2)
-    )
+    ) and
+    getFullyConvertedType(node).getName() = state
   }
 
-  override predicate isSink(DataFlow::Node node) {
-    exists(PointerAddExpr pae | pae.getAnOperand() = node.asExpr()) or
-    exists(ArrayExpr ae | ae.getArrayBase() = node.asExpr())
+  override predicate isSink(DataFlow::Node node, DataFlow::FlowState state) {
+    (
+      exists(PointerAddExpr pae | pae.getAnOperand() = node.asExpr()) or
+      exists(ArrayExpr ae | ae.getArrayBase() = node.asExpr())
+    ) and
+    getFullyConvertedType(node).getName() = state
```

Comment:
So we're saying that the sink has to have the same (fully converted, unspecified) type as the source.  Why does this help performance?  Why does it not affect results?

Reply:
Yes. Let me answer each question separately:
### Why does it not affect results?
Because the query is already requiring that the type of the source and the sink is equivalent. The body of the query is:
```ql
cfg.hasFlowPath(pragma[only_bind_into](source), pragma[only_bind_into](sink)) and
hasFullyConvertedType(source, t) and
hasFullyConvertedType(sink, t)
```

Reply:
Exactly. Let me answer each question separately:
### Why does it not affect results?
Because the query is already requiring that the type of the source and the sink is equivalent. To see this, notice that the body of the `select` is:
```ql
cfg.hasFlowPath(pragma[only_bind_into](source), pragma[only_bind_into](sink)) and
hasFullyConvertedType(source, t) and
hasFullyConvertedType(sink, t)
```

### Why does this help performance?
Consider what would happen without this PR:
1. We would find all the sources
2. We would find all the sinks
3. We compute all the flows between sources and sinks
4. After computing _all_ these flows, the body of the `select` clause restricts the alerts to only those where the source type is equivalent to the sink type.

This is what we see in:
```ql
[2022-11-18 13:50:58] Evaluated non-recursive predicate #select#cpe#1235#ffff@f7d935aq in 2ms (size: 10).
Evaluated relational algebra for predicate #select#cpe#1235#ffff@f7d935aq with tuple counts:
        9754  ~0%    {4} r1 = SCAN DataFlowImpl#9021cc4c::Configuration::hasFlowPath#2#dispred#fff OUTPUT In.1, In.2, In.2, In.1
        9754  ~2%    {5} r2 = JOIN r1 WITH CastArrayPointerArithmetic#b621a59c::hasFullyConvertedType#2#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1, Lhs.0, Lhs.2, Lhs.3
          10  ~0%    {4} r3 = JOIN r2 WITH CastArrayPointerArithmetic#b621a59c::hasFullyConvertedType#2#ff ON FIRST 2 OUTPUT Lhs.0, Lhs.2, Lhs.3, Lhs.4
                     return r3
```
There are `9754` pairs of sources and sinks, but only `10` of those end up being reported because their types match.

Now, consider what happens with this PR:
1. We would find all the sources (as before)
2. We would find all the sinks (as before)
3. We compute all the flows between sources and sinks (as before). But now there's only a flow between a source and a sink when they have the same (string representation) of their type. This means there are _far_ fewer paths.
4. After computing these flows, the body of the `select` clause restricts the alerts to only those where the source type is equivalent to the sink type (as before).

```ql
The important part is step 3. We see the effect of the PR evaluation of the `select` body:
[2022-11-18 14:41:49] Evaluated non-recursive predicate #select#cpe#1235#ffff@d790a1nc in 0ms (size: 10).
Evaluated relational algebra for predicate #select#cpe#1235#ffff@d790a1nc with tuple counts:
        10  ~0%    {4} r1 = SCAN DataFlowImpl#9021cc4c::Configuration::hasFlowPath#2#dispred#fff OUTPUT In.1, In.2, In.2, In.1
        10  ~0%    {5} r2 = JOIN r1 WITH CastArrayPointerArithmetic#b621a59c::hasFullyConvertedType#2#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1, Lhs.0, Lhs.2, Lhs.3
        10  ~0%    {4} r3 = JOIN r2 WITH CastArrayPointerArithmetic#b621a59c::hasFullyConvertedType#2#ff ON FIRST 2 OUTPUT Lhs.0, Lhs.2, Lhs.3, Lhs.4
                   return r3
```

Reply:
> We compute all the flows between sources and sinks (as before). But now there's only a flow between a source and a sink when they have the same (string representation) of their type. This means there are far fewer paths.

I guess my intuition was that dataflow would compute all the same paths as before, but throw them out at the end where the flow states don't match the sinks (i.e. the same as we were doing manually before).  But this clearly isn't the case.  It constructs differently labelled edges and then fails to find flow between the different types, greatly reducing the tree that is ever explored.

Thanks for the explanation, and apologies that some of it appears to be repeated from the PR description.